src/SwiftPackage.ts

```diff
@@ -157,7 +165,8 @@ export class SwiftPackage implements PackageContents {
         readonly folder: vscode.Uri,
         private contents: SwiftPackageState,
         public resolved: PackageResolved | undefined,
-        public plugins: PackagePlugin[]
+        public plugins: PackagePlugin[],
+        public workspaceState: WorkspaceState | undefined
```

Comment:
If you are caching workspaceState there is a good chance it'll get out of date. You can't just reload it when the Package.swift is reloaded. You can't even just reload it when the Package.resolved is reloaded, as it is dependent on a resolve having finished. You could add another FileSystemWatcher to catch when the file changes, but I don't know how often the file is touched by the SwiftPM system. If it is every build then it isn't worth caching.

Reply:
Thanks!  I was thinking about the update timing

based on the elements of `workspace-state.json`, it would be likely not much change very often.
currently, I think they could be 
* `new package added/deleted`
* `edit status switching`

Just checked with adding a new package in `Package.swift`,  the `Package.resolved` and `workspace-state.json` are updated both 

Maybe it's good to add `FileSystemWatcher`

And I changed my package.swift to `5.5`, dependencies can show well.

>  If it is every build then it isn't worth caching.

maybe it's not that expensive?  even it's changed the cached dependencies might be worthy to show for a while (better than emptying them immdediately?)

Reply:
> based on the elements of workspace-state.json, it would be likely not much change very often.
currently, I think they could be

It could change whenever you do a resolve, update, edit, build. Package.resolved is changed by source control.

> Maybe it's good to add FileSystemWatcher

I just did a little test, and every time any action occurs, the workspace-state.json is touched (even though it doesn't change). This kicks off a `FileSystemWatcher` change event.

> maybe it's not that expensive? even it's changed the cached dependencies might be worthy to show for a while (better than emptying them immdediately?)

If it isn't expensive there is no point caching it. The amount of times you need the data is less than the amount of times you build the project.

Given all of the above can we get rid of the caching

Reply:
I just think about it again, updating UI could be more expensive than updating data?  because in certain cases that the file is changed but there is no actual change for data, so we can just compare the data, and once found they are the same then we can ignore the UI-updating?

Reply:
removed